File: src/main/java/com/uwetrottmann/trakt/v2/entities/MovieCheckin.java
package com.uwetrottmann.trakt.v2.entities;

public class MovieCheckin extends BaseCheckin {

    public SyncMovie movie;

    public static class Builder {

        private SyncMovie movie;
        protected ShareSettings sharing;
        protected String message;
        protected String venue_id;
        protected String venue_name;
        protected String app_version;
        protected String app_date;

        public Builder(SyncMovie movie, String appVersion, String appDate) {
            if (movie == null) {
                throw new IllegalArgumentException("Movie must not be null");
            }
            this.movie = movie;
            this.app_version = appVersion;
            this.app_date = appDate;
        }

        public Builder sharing(ShareSettings shareSettings) {
            this.sharing = shareSettings;
            return this;
        }

        public Builder message(String message) {
            this.message = message;
            return this;
        }

        public Builder venueId(String venueId) {
            this.venue_id = venueId;
            return this;
        }

        public Builder venueName(String venueName) {
            this.venue_name = venueName;
            return this;
        }

        public MovieCheckin build() {
            MovieCheckin checkin = new MovieCheckin();
            checkin.movie = movie;
            checkin.sharing = sharing;
            checkin.message = message;
            checkin.venue_id = venue_id;
            checkin.venue_name = venue_name;
            checkin.app_date = app_date;
            checkin.app_version = app_version;
            return checkin;
        }
    }

}

File: src/main/kotlin/com/redissi/trakt/entities/MovieCheckin.kt
package com.redissi.trakt.entities

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
class MovieCheckin(
    override val sharing: ShareSettings? = null,
    override val message: String? = null,
    @Json(name = "venue_id")
    override val venueId: String? = null,
    @Json(name = "venue_name")
    override val venueName: String? = null,
    @Json(name = "app_version")
    override val appVersion: String? = null,
    @Json(name = "app_date")
    override val appDate: String? = null,
    val movie: SyncMovie? = null
) : BaseCheckin() {

    class Builder(
        movie: SyncMovie?,
        appVersion: String,
        appDate: String
    ) {
        private val movie: SyncMovie
        private var sharing: ShareSettings? = null
        private var message: String? = null
        private var venueId: String? = null
        private var venueName: String? = null
        private var appVersion: String
        private var appDate: String

        fun sharing(shareSettings: ShareSettings?): Builder {
            sharing = shareSettings
            return this
        }

        fun message(message: String?): Builder {
            this.message = message
            return this
        }

        fun venueId(venueId: String?): Builder {
            this.venueId = venueId
            return this
        }

        fun venueName(venueName: String?): Builder {
            this.venueName = venueName
            return this
        }

        fun build(): MovieCheckin {
            return MovieCheckin(
                movie = movie,
                sharing = sharing,
                message = message,
                venueId = venueId,
                venueName = venueName,
                appDate = appDate,
                appVersion = appVersion
            )
        }

        init {
            requireNotNull(movie) { "Movie must not be null" }
            this.movie = movie
            this.appVersion = appVersion
            this.appDate = appDate
        }
    }
}